rfc-141-application-healthchecks.md

```diff
@@ -0,0 +1,225 @@
+
+- For load balancing, to determine if an instance is ready to serve
+  requests.
+- For continuous deployments, to determine (along with other smoke
+  tests) if a release can be automatically pushed to production.
+- For alerting, so that 2ndline know if an instance has a problem
+  which needs manual intervention to fix.
+
+We implement healthchecks using the [`GovukHealthcheck`][] module in
+[govuk_app_config][].
+
+Currently, a healthcheck response is always served with an HTTP status
+of 200, and indicates in the body whether the instance's status is
+"ok", "warning", or "critical".  However, AWS load balancers use the
+HTTP status, not the response body, to determine whether an instance
+is healthy.  So we will continue to send requests to an instance which
+is in a "warning" or "critical" state.
+
+This RFC proposes that we standardise on our healthchecks returning a
+200 or a 500 status code, which has some further ramifications on
+monitoring and alerting (discussed below).
+
+A full list of healthchecks which will need updating is given at the
+end of the document.
+
+## Problem
+
+We overload the meaning of "healthcheck", and cover both *application
+health* as well as *instance health* with our `/healthcheck`
+endpoints.  This introduces confusion, and means that we cannot
+currently use healthchecks in load balancing.
+
+Here are some examples of healthchecks which are suitable for load
+balancing and for post-deployment checks:
+
+- [`GovukHealthcheck::ActiveRecord`][]
+- [`GovukHealthcheck::Mongoid`][]
+- [`GovukHealthcheck::RailsCache`][]
+- [`GovukHealthcheck::Redis`][]
+- [`GovukHealthcheck::SidekiqRedis`][]
+
+If an app can't talk to a backing service it relies on, then it
+probably can't be trusted to handle any requests.  One of these
+healthchecks failing could indicate a networking or a configuration
+issue.
+
+Here are some examples of healthchecks which are unsuitable for load
+balancing or for post-deployment checks:
+
+- [`GovukHealthcheck::SidekiqQueueCheck`][], this is an abstract check
+  which other checks extend.
+  [`GovukHealthcheck::SidekiqQueueLatencyCheck`][] is an instance of
+  this check.
+- [`GovukHealthcheck::ThresholdCheck`][], this is another abstract
+  check which other checks extend.
+  [`GovukHealthcheck::SidekiqRetrySizeCheck`][] is an instance of this
+  check.
+- [`Healthcheck::ApiTokens`][] in [signon][], which reports that an
+  API token needs rotating.
+
+If one of these healthchecks fail, the app instances themselves are
+probably fine.  The problem is elsewhere.
+
+Two of these indicate capacity problems outside of the instance, and
+the third is an alert that a manual maintenance procedure needs to be
+performed some time in the next two months.  They all share the
+property that if one instance reports a failure, all will.  And so if
+we use these healthchecks as part of load balancing rules, all running
+instances of the app will be taken down simultaneously.
```

Comment:
This is true for the few remaining scary Classic ELB load balancers (and unfortunately I think there's at least one of these handling actual traffic). Well-designed load balancers like ALBs (which most of ours are) don't do this and will spray to all backends when all backends are simultaneously unhealthy, rather than giving up and just dropping traffic on the floor. Still bad though, because the healthcheck then becomes ineffective and an actually-failed backend will still get traffic.

I'd probably simplify this to something like

```suggestion
property that if one instance reports a failure, all will.  This makes
them unsuitable for load balancing purposes.```

(Or if you wanted to include all the detail, maybe "Depending on the load balancer, either all traffic will be dropped or the healthcheck will become ineffective and traffic will be routed even to instances which have actually failed.")